### PR TITLE
ARTEMIS-1461 Allow changing of messages through interceptors

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

**Comment by tabish121** [comment]
So I don't see any documentation explaining how this violates the AMQP specification about the immutability of the bare message part of the AMQP message, or how it can now can present a message to the receiver with the same MessageId that is not in fact the message that was sent.  Also it would break anyone that uses message signing as the original Footer would likely be preserved and so again present to the receiver a message pretending to be the one that was sent but in fact not be and thereby not match any included signature.

**Comment by clebertsuconic** [comment]
Users can do that already today.  Not introducing anything new.  


Not recommended. But users can do that already.

**Comment by clebertsuconic** [comment]
We would need to change putProperties to forbid it. 


All we can do is document it.

**Comment by clebertsuconic** [comment]
@tabish121 I have changed docs around to address your point

**Comment by harrison-tarr** [comment]
I pulled this down and tested it with my ServerPlugin and AMQP producer/consumer code. I'm no longer seeing the issue I was before, so I think this fixes it 🎉

**Comment by clebertsuconic** [comment]
@harrison-tarr mind reviewing the doc I wrote?

File: examples/features/standard/broker-plugin/src/main/java/org/apache/activemq/artemis/jms/example/BrokerPlugin.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+   int count = 0;
+
+   @Override
+   public void beforeSend(ServerSession session,
+                          Transaction tx,
+                          Message message, boolean direct,
+                          boolean noAutoCreateQueue) throws ActiveMQException {
+      if (message instanceof AMQPMessage) {
+         // AMQP Messages are by definition immutable.
```
Comment (harrison-tarr) [inline examples/features/standard/broker-plugin/src/main/java/org/apache/activemq/artemis/jms/example/BrokerPlugin.java]: I don't think you need to treat these differently based on the type of message. I think you can just say 
`message.putStringProperty()
message.reencode()`
If i remember correctly, ICoreMessage just inherits the default noop implementation of `reencode`

**Comment by harrison-tarr** [comment]
@clebertsuconic it looks good to me. It might be worth adding a note in the https://activemq.apache.org/artemis/docs/latest/broker-plugins.html document about this example

File: examples/features/standard/broker-plugin/src/main/java/org/apache/activemq/artemis/jms/example/BrokerPlugin.java (hunk 1 above)
Comment (clebertsuconic) [reply examples/features/standard/broker-plugin/src/main/java/org/apache/activemq/artemis/jms/example/BrokerPlugin.java]: @harrison-tarr  that's right... and do you believe I wrote that? :)

**Comment by Matze2** [comment]
I tried the latest snapshot with an AmqpInterceptor implementation. This does not work.
```
	@Override
	public boolean intercept(final AMQPMessage message, RemotingConnection connection) {
		logger.info("AMQP Interceptor gets called with message " + message.getMessageID());
		logger.info("Properties before: " + message.getPropertyNames());
		message.putStringProperty("newProp", "newPropValue");
		logger.info("Properties after: " + message.getPropertyNames());
		return true;
	}

```
The listener does not see the "newProp" property added here.
Will try the broker plugin example now.

**Comment by Matze2** [comment]
Ignore my last message. I was irritated by the pull request title containing the word "interceptor". The JIRA issue is named differently.

**Comment by clebertsuconic** [comment]
I think the snapshot wasn’t updated yet.

**Comment by Matze2** [comment]
I was not even aware that there is somewhere a binary snapshot to download.
I compiled it myself (had to add servlet-api dependency to artemis-rest module to be able to build it).

And I now tested the broker plugin case and this works. Thanks!

**Comment by clebertsuconic** [comment]
@Matze2 the broker plugin is basically intercepting the message. I used the term out of the concept instead of the class name.

**Pull request closed without merging**